### ISSUE TITLE
Usage metrics (module-wise)

### ISSUE THREAD

**Issue by swapagarwal** [swapagarwal/JARVIS-on-Messenger#247]
Think of a good way to gather usage metrics, and analyze them; e.g. which modules are being used the most, how many times a module is consecutively used, what's the correlation between two modules, etc.

**Comment by edadesd** [comment]
If someone knows of an applicable library, that would probably be best. Failing that, I could see a database that records username, module called, search term, and time whenever someone uses a module.

**Comment by horrormyth** [comment]
How about deciding a proper convention for having the status or event for each module and use Python logger to record these.Whenever a new module is written, it should strictly reflect the predefined convention.
**For example**
```python
logging.info('Log message',
    extra = {
        'event': 'module_name.called',
        'requester': user_identifier,
        'data': input_data,
        'time':timezone.now(),

    }
)
```
and same for the errors and exception as well. Having these we will have proper events to track.

**Comment by swapagarwal** [comment]
@edadesd How about Firebase for storing the data? I'm not sure how easy is it to query.
@horrormyth That is the ideal way to go! We need to decide on the fields that are relevant, and then decide the logging schema.

**Comment by horrormyth** [comment]
Yup do you have anything in mind regarding the scheme and the fields @swapagarwal If we come up with something concrete on this, then this would pave a way to the dashboard feature as well. Regarding the Firebase, what is your take on Firebase as you suggested?

**Comment by edadesd** [comment]
It looks like there are some Python libraries for Firebase: https://firebase.google.com/docs/database/rest/start

I like the idea of using the Python logging library. It looks like it can send pertinent information to a log file with every call, which sounds useful to build a database out of.

**Comment by horrormyth** [comment]
Yes, @edadesd  that would be the simplest and ideal way to catch every event, however, the proper schema, the model and DB should be decided.As long as these are ready, I hope the rest should be not trivial.

**Comment by horrormyth** [comment]
is there any update on this? @swapagarwal, @edadesd

**Comment by edadesd** [comment]
I don't have anything to add right now.

**Comment by swapagarwal** [comment]
I thought about the fields required in schema, and came up with the following:
input
success
output
intent
entities
user
module
timestamp
@edadesd @horrormyth What do you think?

**Comment by swapagarwal** [comment]
I'm exploring 2 options:
* Integrate with a plugin in heroku (that supports Kibana)
* Integrate with AWS (something like Athena)

**Comment by edadesd** [comment]
If the module has quick replies, we might want to record whether a user clicked one, and which.

**Comment by swapagarwal** [comment]
Also, `postback` data should be logged too.

**Comment by horrormyth** [comment]
@swapagarwal the fields seem okay for me, how abt @edadesd

**Comment by edadesd** [comment]
Yeah, they look good. I think that data can tell the story of what the user wanted the bot to do and whether they got it.

**Comment by edadesd** [comment]
Is it possible to make the logger write directly to a database? Or would we need to process the log into a database using a separate module?

**Comment by horrormyth** [comment]
Sorry for the late comeback :P @edadesd  yes we can set the logger in such a way that it can write straight to the database.However, have we decided yet what are we using for the DB or any other options? I would like to get hands dirty on it as long as we are concrete on the arch.

**Comment by swapagarwal** [comment]
https://github.com/swapagarwal/JARVIS-on-Messenger/milestone/2

**Comment by hiteshramani** [comment]
@swapagarwal Not sure if this requirement is taken care of. If not, have you thought of using New Relic?
